src/gpl/src/cudasplibs.cu

```diff
@@ -0,0 +1,119 @@
+    cudaerror(cudaMemcpy(d_cooVal, cooVal.data(), sizeof(float)*nnz, cudaMemcpyHostToDevice));
+    cudaerror(cudaMemcpy(d_fixedInstForceVec, fixedInstForceVec.data(), sizeof(float)*m, cudaMemcpyHostToDevice));
+}
+
+void cudasplibs::cusolverSpQR(Eigen::VectorXf& instLocVec){
+    // Parameters that don't change with iteration and used in the CUDA code
+    float tol = 1e-6; // 	Tolerance to decide if singular or not.
+    int reorder = 0;  // "0" for common matrix without ordering
+    int singularity = -1; // Output. -1 = A means invertible
+    
+    // Set handler
+    cusolverSpHandle_t handleCusolver = NULL;
+    cusparseHandle_t handleCusparse = NULL;
+    cudaStream_t streamX = NULL;
+    
+    // Initialize handler
+    cusolvererror(cusolverSpCreate(&handleCusolver));
+    cusparseerror(cusparseCreate(&handleCusparse));
+    cudaerror(cudaStreamCreate(&streamX));
+    cusolvererror(cusolverSpSetStream(handleCusolver, streamX));
+    cusparseerror(cusparseSetStream(handleCusparse, streamX));
+
+    // Create and define cusparse descriptor
+    cusparseMatDescr_t descr = NULL;
+    cusparseerror(cusparseCreateMatDescr(&descr));
+    cusparseerror(cusparseSetMatType(descr, CUSPARSE_MATRIX_TYPE_GENERAL));
+    cusparseerror(cusparseSetMatIndexBase(descr, CUSPARSE_INDEX_BASE_ZERO));
+
+    // transform from coordinates (COO) values to compressed row pointers (CSR) values
+    // https://docs.nvidia.com/cuda/cusparse/index.html
+    int *d_csrRowInd = NULL;
+    cudaerror(cudaMalloc((void**)&d_csrRowInd, (m+1) * sizeof(int)));
+    cusparseerror(cusparseXcoo2csr(handleCusparse, d_cooRowIndex, nnz, m, d_csrRowInd, CUSPARSE_INDEX_BASE_ZERO));
+    cusolvererror(cusolverSpScsrlsvqr(handleCusolver, m, nnz, descr, d_cooVal, d_csrRowInd, d_cooColIndex, d_fixedInstForceVec, tol, reorder, d_instLocVec, &singularity));
+
+    // Sync and Copy data to host
+    cudaerror(cudaMemcpyAsync(instLocVec.data(), d_instLocVec, sizeof(float)*m, cudaMemcpyDeviceToHost, streamX));
+
+    cudaerror(cudaFree(d_csrRowInd));
+    cusparseerror(cusparseDestroyMatDescr( descr ) );
+    cusparseerror(cusparseDestroy(handleCusparse));
+    cusolvererror(cusolverSpDestroy(handleCusolver));  
+}
+
+__global__ void Multi_MatVec(float error, int nnz, int m , float* d_Ax, float* d_fixedInstForceVec, float* d_instLocVec, int* d_cooRowIndex, int* d_cooColIndex, float* d_cooVal){
+    float sum = 0;
+    int num = blockIdx.x * blockDim.x + threadIdx. x;
+    if (num < nnz){
+        d_Ax[d_cooRowIndex[num]] += d_cooVal[num] * d_instLocVec[d_cooColIndex[num]];
+    }
+    for (size_t row = 0; row < m; row++){
+        sum += (d_fixedInstForceVec[row] > 0) ? d_fixedInstForceVec[row] : -d_fixedInstForceVec[row];
+        if (d_fixedInstForceVec[row] > d_Ax[row])
+            error += d_fixedInstForceVec[row] - d_Ax[row];
+        else
+            error -= d_fixedInstForceVec[row]- d_Ax[row];
+    }
+    if (sum != 0) error = error / sum;
+}
+
+float cudasplibs::error_cal(){
```

Comment:
Can .cu files be clang-formatted?

Reply:
I used clang-format for all files I added.

Reply:
clearly you didn't here: "cudasplibs::error_cal(){" is not formatted.

Reply:
Sorry I will do it and check every place else.

Reply:
Oh. I did the wrong explanation before. Are you checking with the newest version? I have deleted "cudasplibs.cu". I am using "solver.h" and "gpuSolver.h" currently.